core/shared/platform/windows/win_thread.c

```diff
@@ -332,6 +390,7 @@ os_thread_join(korp_tid thread, void **p_retval)
     os_sem_wait(&curr_thread_data->wait_node.sem);
     if (p_retval)
         *p_retval = curr_thread_data->wait_node.retval;
+    thread_data_list_remove(thread_data);
```

Comment:
If `thread_data_list_remove()` thread_data now, what will happen if caller join this thread again? Can `os_thread_join()` figure out the thread has already terminated and return.

---
https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread.join?view=net-9.0
>If the thread has already terminated when [Join](https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread.join?view=net-9.0) is called, the method returns immediately

Reply:
Other limitation include:

> You should never call the [Join](https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread.join?view=net-9.0) method of the [Thread](https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread?view=net-9.0) object that represents the current thread from the current thread. This causes your app to become unresponsive because the current thread waits upon itself indefinitely,

> You cannot invoke Join on a thread that is in the [ThreadState.Unstarted](https://learn.microsoft.com/en-us/dotnet/api/system.threading.threadstate?view=net-9.0#system-threading-threadstate-unstarted) state.

Reply:
> If thread_data_list_remove() thread_data now, what will happen if caller join this thread again? Can os_thread_join() figure out the thread has already terminated and return.

When it tries to join, L360-363 will return `BHT_ERROR`

> You should never call the [Join](https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread.join?view=net-9.0) method of the [Thread](https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread?view=net-9.0) object that represents the current thread from the current thread. This causes your app to become unresponsive because the current thread waits upon itself indefinitely

It will deadlock too, L390 sem wait forever

> You cannot invoke Join on a thread that is in the [ThreadState.Unstarted](https://learn.microsoft.com/en-us/dotnet/api/system.threading.threadstate?view=net-9.0#system-threading-threadstate-unstarted) state.

Our implementation uses `_beginthread` and didn't use Thread object, won't have this issue